designs/EmbeddedReadmesTechSpec.md

```diff
@@ -0,0 +1,88 @@
+
+```
+Readme="$(PackageReadmeFile)"
+```
+
+Further, a `GetReadme` method must be added to the [NuspecReader.cs](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Packaging/NuspecReader.cs#L20) file to facilitate the reading of the Readme element from a nuspec file.
+
+Next, we need to add `Readme` member variables to multiple relevant classes and interfaces, and ensure that the Readme property is transmitted from one class to another when `pack` is called. These are:
+* [IPackTaskRequest](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Build.Tasks.Pack/IPackTaskRequest.cs#L16)
+* [PackTask](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTask.cs#L14)
+* [PackTaskRequest](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskRequest.cs#L10)
+* [IPackageMetadata](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/IPackageMetadata.cs#L11)
+* [ManifestMetadata](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/ManifestMetadata.cs#L18)
+* [PackageBuilder](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L25)
+
+In order to transmit the `Readme` value from the source file (*nuspec* or *csproj*) to the final output *nuspec* file, the appropriate assignment statements are needed at the relevant locations.
+
+After this, a `ManifestMetadata` object is used to create the output *nuspec* file that is included in the nupkg. For this step, we need to edit the [ToXElement](https://github.com/NuGet/NuGet.Client/blob/fa61e76d296b4b37ef4226277e77f7f227e878d9/src/NuGet.Core/NuGet.Packaging/PackageCreation/Xml/PackageMetadataXmlExtensions.cs#L32) method in PackageMetadataXmlExtensions.cs to include this line:
+
+```
+AddElementIfNotNull(elem, ns, "readme", metadata.Readme);
+```
+
+**Validation and Errors**
+
+To validate the Readme property and the corresponding file when `pack` is called, the ValidateReadmeFile method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) checks that:
+* The file extension matches a Markdown file
+* The file specified in the Readme property exists in the package's files
+* The file size does not exceed 1 MB
```

Comment:
These validations are mainly for nuget.org, right? If we eventually want to have looser restrictions for readmes for local or private feeds, we feel this is a minimum set of validations going forward? eg, In future iterations of this feature, I could have a 5GB readme on my private feed... who cares? But I can't publish to nuget.org with that. 

I think the answer here is "yes", so may want to mention that in the description how we feel this validation set could evolve without causing backwards compatibility issues.

If you don't like the idea of having Pack validation eventually removed/less restrictive, then I'd question why we need to validate on Pack and not on Push?...
```
Pack -> warn that nuget.org won't accept >1MB
Push -> error from nuget.org (server doesn't accept > 1MB)
(in the future)
Push  (NOT nuget.org) -> accept/error depending on server behavior
Push (local feed) -> local feed doesn't care
```

Reply:
@chgill-MSFT I was hoping you could clarify this

Reply:
This is a good question!

Advay and I had a chat about this. It looks like we don't validate for size on license file today. In general we don't make any assumptions on pack that a package isn't meant for NuGet.org today. Not only that, but the likelihood that a README will be larger than 1mb is very low.

Ultimately, we came to the conclusion that is makes sense to remove the size check on pack and only validate README file size on push/upload to NuGet.org.

Reply:
Sounds good! I think keeps our options open in the long run. 
Are we planning on changing `pack` to `push` in this paragraph?

Reply:
Nevermind, yes the basic checks will remain on `pack`. I'm guessing `push` validations are out of scope? May want to indicate that in **Future**?

Reply:
Since we won't be doing the validation on our side and the size check will only be done by nuget.org, should we say we validate size on `push`, or should we just add a note saying that "NuGet.org will only accept README files that do not exceed 1 MB"?

Reply:
I'll add the 1 MB size validation to the "Considerations" section and explain why we won't be doing it here. I'll include a reference to NuGet.org's 1 MB size limit there, and mention that this would result in an error when someone uses `push` to upload a package to NuGet.org. Does that seem fair?

Reply:
Done!